New article rules engine (#521)
akeneo/pim-helpcenter: content/md/productivity/get-started-with-the-rules-engine.md

---
id: get-started-with-the-rules-engine
themes: boost-your-productivity, catalog-settings
title: Get started with the **rules engine**
popular: false
ee-only: true
related: what-is-a-rule
---

The rules engine can radically boost your productivity in the PIM by automating actions. The rules are defined in a YML file that you can import in the PIM, using a dedicated job profile.   

> This feature is very, like, really powerful, but it can also be a bit complex to apprehend. Reading this article, you will find out what you can use it for and how it works.

Reviewer: I'd remove the "complex to apprehend" to turn it in a positive way.

Author: What do you propose?

> 

Reviewer 2: I would add a paragraph about the general behavior: a rule is composed of 1 or several conditions to target a list of products and applicates 1 or several actions to this list.

Reviewer 2: And there is not a full example of a rule. Just a simple example of YML file with 1 condition and 1  action. 
(If we can take the top use case from our customers it will be perfect!)

# Available actions

As of today, 5 actions are available in the rules engine:
- copy
- add
- set
- remove
- concatenate

What are these actions for? Let's discover it :wink:

## Copy

The `Copy` action enables you to copy an attribute value into another one.

The expected values are:
 - `from_field`: the code of the attribute to be copied
 - `to_field`: the attribute code the value will be copied into
 - `from_locale`: the locale code of the value to be copied *(optional)*
 - `from_scope`: the channel code of the value to be copied *(optional)*
 - `to_locale`: the locale code the value will be copied into *(optional)*
 - `to_scope`: the channel code the value will be copied into *(optional)*

### Example

You have a *scopable* and *localizable* attribute called `description`. You can copy its content from the `en_US` locale and the `print` channel to the `en_US` locale and the `e-commerce` channel. The action will be defined as follows:

```YML
  actions:
      type: copy
        from_field: description
        from_locale: en_US
        from_scope: print
        to_field: description
        to_locale: en_US
        to_scope: ecommerce

> ```

Reviewer 2: Actions represent a list of actions. So this is an array (represents by a `-` in YML).  
And becareful, the line `type: copy` must be at same level as the from_field, etc... 

```suggestion
```YML
  actions:
      -
        type: copy
        from_field: description
        from_locale: en_US
        from_scope: print
        to_field: description
        to_locale: en_US
        to_scope: ecommerce
```

```

## Add

> The `Add` action allows you to add values to a multi-select attribute or a reference entity multiple link attribute. Also, it makes it possible to add a product to categories.

Reviewer 3: to association and to groups too !

The expected values are:

> - `field`: the attribute code

Reviewer 3: the "field" code, as it can be an attribute, category, association or group.

> - `value`: the attribute value

Reviewer 3: Same here.

Reviewer 3: You wrote "value", and the example is "items". Say that it has to be an array of the items to add.

Author: To check together, not sure to understand what to write exactly.

- `locale`: the locale code for which the value is assigned (optional)
- `scope`: the channel code for which the value is assigned (optional)

#### Example

> To add the “t-shirts” category, the action will be as follows:

Reviewer 3: ```suggestion
To add the “t-shirts” category in a set of products, the action will be as follows:
```

```YML
  actions:
      type: add
        field: categories
        items:
          - t-shirts

> ```

Reviewer 2: same remark (and for others too)

Author: Sorry Nini, what are talking about?

### Set

The `Set` action assigns value(s) to every attributes, categories, family, associations, groups, the "parent" field, or the "enabled" field.

The expected values are:

> - `field`: the attribute code or category code or family code

Reviewer 3: field (sans under)

- `value`: the attribute value or category code or family code
- `locale`: the locale code for which the value is assigned (optional)
- `scope`: the channel code for which the value is assigned (optional)

#### Example

To set the *“My very new description for purple T-shirt”* value to your `description` attribute in the `en_US` locale and for the `e-commerce` channel, the action will be as follows:

```YML
  actions:
      type:   set
        field:  description
        locale: en_US
        scope:  ecommerce
        value:  "My very new description for purple tshirt"
```

## Remove

> The `Remove` action enables you to remove values from a multi-select attribute, a reference entity multiple link attribute or a product category.

Reviewer 3: or groups too.

The expected values are:

> - `field`: the attribute code or “categories”

Reviewer 3: field, be consistent with the other ones

- `items`: the values to remove
- `locale`: the locale code for which the value is assigned (optional)
- `scope`: the channel code for which the value is assigned (optional)
- `include_children`: if true, then it also applies the removal of the children to the given categories. It is only applicable if the field is set to “categories” (It's optional. By default, it is set to false)

### Example

To remove the “t-shirts” category, the action will be as follows:

```YML
  actions:
      type: remove
        field: categories
        items:
          - t-shirts
```

To remove the “clothing” category and its children, the action will be as follows:
```YML
  actions:
      type: remove
      field: categories
      items:
        - clothing
      include_children: true
```

To unclassify products from the whole “Master catalog” tree, the action will be as follows:

```YML
  actions:
      type: remove
      field: categories
      items:
        - master
      include_children: true
```

## Concatenate

> The `Concatenate` action concatenates at least two values into a single value. A space separates each source value.

Reviewer: what about adding in the intro that it only concerns attribute vales?

The possible source attribute types are:
 - text
 - text area
 - date
 - identifier
 - metric
 - number
 - price collection
 - simple or multi select (option codes)

The possible target attribute types are:
 - text

>  - text area

Reviewer 3: I think it's "textarea"

Author: in the UI, it's "text area". Don't know if it's better to use the UI wording or the technical one. Same for "price". In the UI, it's not "price collection", but "price".

**The parameters `from` and `to` are required in this format. Depending on the source attribute type, some optional keys can be added.**

The expected values are:

**`from`**
- `field`: the attribute code.
- `locale`: the locale code for which the value is assigned (optional)
- `scope`: the channel code for which the value is assigned (optional)
- For `date` attribute: the format of the date following the PHP format [specification](https://www.php.net/manual/en/function.date.php) (Optional). By default, it is *Y-m-d* (e.g. *2020-01-31*)
- For `price collection` attribute: the currency code for which the price is assigned (Optional). By default all the prices in the collection are displayed, separated by a coma.

**`to`**
- `field`: the attribute code.
- `locale`: the locale code for which the value is assigned (optional)
- `scope`: the channel code for which the value is assigned (optional)

### Example

> To concatenate the **brand** (non localizable and non scopable) and the **model** in the `en_US` locale into the **description** value in the `en_US` locale, the action will be as follows:

Reviewer: Brand & model into a Product title would be a better match I guess. Unless you're targeting short description ;)

```YML
  actions:
    - type: concatenate

>         from:

Reviewer 3: You forget 2 "-"

            -
                field: brand
            - 
                field: model
                locale: en_US

            field: brand
            field: model
            locale: en_US
        to:
            field: description

>             locale: en_US

Reviewer 2: ```suggestion
    - type: concatenate
      from:
          field: brand
          field: model
          locale: en_US
      to:
          field: description
          locale: en_US
```
or 

```suggestion
    - 
      type: concatenate
      from:
          field: brand
          field: model
          locale: en_US
      to:
          field: description
          locale: en_US
```

```

To concatenate the **model** in the `en_US` locale, the **color** in the `en_US` locale and the **year of the release date** into the **title** value in the `en_US` locale, the action will be as follows:

```YML
  actions:
    - type: concatenate
        from:

>             field: model

Reviewer 3: Same here, you forget '-'

            locale: en_US
            field: color
            locale: en_US
            field: release_date
            format: Y
        to:
            field: title
            locale: en_US
```

> To concatenate the **model** in the `en_US` locale and the **price** in USD from the `mobile` channel into the **subtitle** value in the `en_US` locale and the `mobile` channel, the action will be as follows:

Reviewer: Examples look weird to me. Do you feel they match customer use cases?

Author: honestly, I didn't change the examples :D I copy pasted the content of nini. But yes, I'll change to make it more user friendly

Author: Actually we can remove this example, we already have 2 examples for this action, whereas all the other ones have only 1 example. Maybe we can remove 2 of 3 examples. What do you think?

```YML
  actions:
    - type: concatenate
        from:

>           - field: model

Reviewer 3: This one is right :)

            locale: en_US
          - field: price
            scope: mobile
            currency: USD
        to:
            field: subtitle
            locale: en_US
            scope: mobile
```

> Now that you have discovered all the available actions, it's important to have in mind all the **fields** and their **operators** that we support in the rules engine. The "fields" are the different product properties on which you can filter (in other words, it defines your **selection**).

Reviewer 3: Didn't understand the first sentence

# Available fields

As of today, 7 fields are supported in the rules engine, and each of them has it's own business rules. Here we go!

## Created

The possible operators for the `created` field are:
- =
- !=
- “>”
- <
- BETWEEN
- NOT BETWEEN
- EMPTY
- NOT EMPTY

::: info
The format of the date is: yyyy-mm-dd.
:::

:::info

> If the operator is EMPTY or NOT EMPTY, the value element will be ignored.

Reviewer: maybe elaborate on the "value element will be ignored"

## Text/Text area

The possible operators for the `text/text area` attribute types are:  

- STARTS WITH
- ENDS WITH
- CONTAINS
- DOES NOT CONTAIN
- =
- !=
- EMPTY
- NOT EMPTY

:::info
If the operator is EMPTY or NOT EMPTY, the value element will be ignored.
:::

### Example

```YML
field: description
operator: CONTAINS
value: "Awesome product"
```
**text** can be written with or without quotation marks.

## Metric

The possible operators for the `metric` attribute type are:  
- <
- <=
- =
- !=
- “>”
- “>=”
- EMPTY
- NOT EMPTY

:::info
If the operator is EMPTY or NOT EMPTY, the value element will be ignored.
:::

### Example

```YML
field: weight
operator: =
value:
 amount: 0.5
 unit: KILOGRAM
```

We expect **numeric value** and **measurement unit code** in the following format:
- dot “.” is the decimal separator,
- there is no space between thousands.

## Boolean

The possible operators for the `boolean` attribute type are:  
- >=
- !=

### Example

```YML
field: shippable_us

> operator: =

Reviewer 4: The example doesn't match with the available operators

Author: wow thanks!!! I did a mistake in the operators list ;)

---
id: get-started-with-the-rules-engine
themes: boost-your-productivity, catalog-settings
title: Get started with the **rules engine**
popular: false
ee-only: true
related: what-is-a-rule
---

> The rules engine can radically boost your productivity in the PIM by automating actions. As explained in [this article](what-is-a-rule.html), a rule enables you to automatically do **one or several actions** to a specific **list of products and product models**. To set the targeted products, you have to define `conditions` (that are composed by fields and operators).  

Reviewer 5: ```suggestion
The rules engine can radically boost your productivity in the PIM by automating actions. As explained in [this article](what-is-a-rule.html), a rule enables you to automatically apply **one or several actions** to a specific **list of products and product models**. To set the targeted products, you have to define `conditions` (that are composed by fields and operators).  
```

The rules are defined in a YML file that you can import in the PIM, using a dedicated job profile.   

> This feature is very, like, really powerful, and the usages are large. In order to guide you in the rules engine usage, we wrote this article where you will find out what you can use it for and how it works.

Reviewer 5: ```suggestion
This feature is very, like, really powerful, and the usecases are countless. In order to guide you in the rules engine usage, we wrote this article where you will find out what you can use it for and how it works.
```

# Available actions

As of today, 5 actions are available in the rules engine:
- copy
- add
- set
- remove
- concatenate

What are these actions for? Let's discover it :wink:

## Copy

> The `Copy` action enables you to copy an attribute value into another one.

Reviewer 5: ```suggestion
The `Copy` action enables you to copy an attribute value in another one.
```

The expected values are:
 - `from_field`: the code of the attribute to be copied

>  - `to_field`: the attribute code the value will be copied into

Reviewer 5: ```suggestion
 - `to_field`: the attribute code where the value will be copied
```

 - `from_locale`: the locale code of the value to be copied *(optional)*
 - `from_scope`: the channel code of the value to be copied *(optional)*

>  - `to_locale`: the locale code the value will be copied into *(optional)*

Reviewer 5: ```suggestion
 - `to_locale`: the locale code where the value will be copied *(optional)*
```

>  - `to_scope`: the channel code the value will be copied into *(optional)*

Reviewer 5: ```suggestion
 - `to_scope`: the channel code where the value will be copied *(optional)*
```

You can only copy the given attribute to a selection of attribute types that we defined based on logical criteria:
#### You can copy the option code of a **Simple select** attribute to:
- a **reference entity single link** attribute *(the record should exist)*
- a **text** attribute
- a **textarea** attribute

> #### You can copy the option codes of a **Multi select** attribute to:

Reviewer 5: ```suggestion
#### You can copy the option codes of a **multi select** attribute to:
```

- a **reference entity multiple link** attribute *(the records must already exist)*
- a **text** attribute
- a **textarea** attribute

#### You can copy the value of a **Text** attribute to:
- a **textarea** attribute

> - a **Simple select** attribute (the option code should exist)

Reviewer 5: ```suggestion
- a **simple select** attribute (the option code should exist)
```

- a **Reference entity single link** attribute *(the record should exist)*

#### You can copy the value of an **Identifier** attribute to:
- a **text** attribute
- a **textarea** attribute

#### You can copy the value of a **Date** attribute to:
- a **text** attribute
- a **textarea** attribute

#### You can copy the value of a **Metric** attribute to:
- a **text** attribute
- a **textarea** attribute
- a **number** attribute

#### You can copy the value of a **Number** attribute to:
- a **text** attribute
- a **textarea** attribute
- a **metric** attribute

#### You can copy the value of a **Price** attribute to:
- a **text** attribute
- a **textarea** attribute

#### You can copy the value of a **Reference entity single link** attribute to:
- a **text** attribute
- a **textarea** attribute
- a **simple select** attribute *(the option code of the simple select attribute should exist)*

#### You can copy the value of a **Reference entity multiple link** attribute to:
- a **text** attribute
- a **textarea** attribute

> - a **multi select** attribute *(the option codes of the multi select attribute should exist)*

Reviewer 5: ```suggestion
- a **multi select** attribute *(the option codes of the multi select attribute must already exist)*
```

### Example

You have a *scopable* and *localizable* attribute called `description`. You can copy its content from the `en_US` locale and the `print` channel to the `en_US` locale and the `e-commerce` channel. The action will be defined as follows:

```YML
  actions:
      - type: copy
          from_field: description
          from_locale: en_US
          from_scope: print
          to_field: description
          to_locale: en_US
          to_scope: ecommerce
```

## Add

> The `Add` action allows you to add values to a multi-select attribute or a reference entity multiple link attribute. Also, it makes it possible to add a product to categories, to association and to groups too.

Reviewer 5: ```suggestion
The `Add` action allows you to add values to a multi-select attribute or a reference entity multiple link attribute. It also makes it possible to add a product to categories, to associations, and to groups.
```

The expected values are:
- `field`: the attribute code or property

> - `items`: the value codes. It has to be an array of the items to add.

Reviewer 5: ```suggestion
- `items`: the value codes. It has to be an array of the items you need to add.
```

The expected values are:
- `field`: the attribute code or property
- `value`: the attribute value or property value
- `locale`: the locale code for which the value is assigned (optional)
- `scope`: the channel code for which the value is assigned (optional)

#### Example

To set the *“My very new description for purple T-shirt”* value to your `description` attribute in the `en_US` locale and for the `e-commerce` channel, the action will be as follows:

```YML
  actions:
      - type:   set
          field:  description
          locale: en_US
          scope:  ecommerce
          value:  "My very new description for purple tshirt"
```

## Remove

The `Remove` action enables you to remove values from a multi-select attribute, a reference entity multiple link attribute, a product category or a group.

The expected values are:
- `field`: the attribute code or “categories”
- `items`: the value codes to remove
- `locale`: the locale code for which the value is assigned (optional)
- `scope`: the channel code for which the value is assigned (optional)

> - `include_children`: if true, then it also applies the removal of the children to the given categories. It is only applicable if the field is set to “categories” (It's optional. By default, it is set to false)

Reviewer 5: ```suggestion
- `include_children`: if true, then it also applies the children are also removed from the given categories. It is only applicable if the field is set to “categories” (It's optional. By default, it is set to false)
```

Reviewer 5: Doute sur le sens ici !

### Example

To remove the “t-shirts” category, the action will be as follows:

```YML
  actions:
      - type: remove
          field: categories
          items:
            - t-shirts
```

To remove the “clothing” category and its children, the action will be as follows:
```YML
  actions:
      - type: remove
          field: categories
          items:
            - clothing
          include_children: true
```

To unclassify products from the whole “Master catalog” tree, the action will be as follows:

```YML
  actions:
      - type: remove
          field: categories
          items:
            - master
          include_children: true
```

## Concatenate

The `Concatenate` action concatenates at least two values into a single value. A space separates each source value.

:::info

> The action only concerns attribute values :wink:.

Reviewer 5: ```suggestion
The action is only for attribute values :wink:.
```

:::

The possible source attribute types are:
 - text
 - textarea
 - date
 - identifier
 - metric
 - number
 - price collection
 - simple select
 - multi select (values are separated by a comma)

The possible target attribute types are:
 - text
 - textarea

**The parameters `from` and `to` are required in this format. Depending on the source attribute type, some optional keys can be added.**

The expected values are:

**`from`**
- `field`: the attribute code.
- `locale`: the locale code for which the value is assigned (optional)
- `scope`: the channel code for which the value is assigned (optional)

> - For `date` attribute: the format of the date following the PHP format [specification](https://www.php.net/manual/en/function.date.php) (Optional). By default, it is *Y-m-d* (e.g. *2020-01-31*)

Reviewer 5: ```suggestion
- For the `date` attribute: the format of the date following the PHP format [specification](https://www.php.net/manual/en/function.date.php) (Optional). By default, it is *Y-m-d* (e.g. *2020-01-31*)
```

> - For `price collection` attribute: the currency code for which the price is assigned (Optional). By default all the prices in the collection are displayed, separated by a coma.

Reviewer 5: ```suggestion
- For the `price collection` attribute: the currency code for which the price is assigned (Optional). By default, all the prices in the collection are displayed, separated by a comma.
```

> - For `simple` and `multi select` attributes: in *label_locale*, the expected value is the locale code for the label of the option. Optional. By default, the code of the option is used.

Reviewer 5: ```suggestion
- For the `simple` and `multi-select` attributes: in *label_locale*, the expected value is the locale code for the label of the option. Optional. By default, the code of the option is used.
```

- dot “.” is the decimal separator,
- there is no space between thousands.

## Picture or file

The possible operators for the `Picture` or `File` attribute types are:  
- STARTS WITH
- ENDS WITH
- CONTAINS
- DOES NOT CONTAIN
- =
- !=
- EMPTY
- NOT EMPTY

:::info
If the operator is EMPTY or NOT EMPTY, the value element will be ignored.
:::

### Example

```YML
field: small_image
operator: CONTAINS
value: ../../../
 src/PimEnterprise/Bundle/InstallerBundle/Resources/fixtures/icecat_demo/images/AKNTS_PB.jpg
```

We expect a **text** as `value`.

# A real example
You're done with all these specifications!

We hope that you perfectly understand the rules engine operation now. But to make it crystal clear, we thought it would be better to take a real example of a rule :wink:.

Let's say that, among all the attributes of your product form, you have one `price collection` attribute type and a `boolean` attribute type.
In the price collection, you define the price of your product in USD. The boolean attribute is used to define if your product is ready to be sent to your ecommerce platform (if it is set to "yes", it means that the product is ready).

> Manually change the status of the boolean attribute is time consuming and a very repetitive task... But for the rules engine, it's a perfect match!

Reviewer 5: ```suggestion
Manually changing the status of the boolean attribute is time-consuming and a very repetitive task... But for the rules engine, it's a perfect match!
```

The mission of the rules engine here is to **automate the status change of the boolean once the price attribute has a value.**

> For that, follow these steps:

Reviewer 5: ```suggestion
To do so, follow these steps:
```

> 1. Open an editor text to write the YML code of the rule.

Reviewer 5: ```suggestion
1. Open a text editor to write the YML code of the rule.
```

 - `to_locale`: the locale code the value will be copied into *(optional)*
 - `to_scope`: the channel code the value will be copied into *(optional)*

You can only copy the given attribute to a selection of attribute types that we have defined based on logical criteria:
#### You can copy the option code of a **Simple select** attribute to:
- a **reference entity single link** attribute *(the record should exist)*
- a **text** attribute
- a **textarea** attribute

#### You can copy the option codes of a **Multi select** attribute to:
- a **reference entity multiple link** attribute *(the records should exist)*

> - a **text** attribute

Reviewer 2: The codes are separated by something? Maybe it could be useful to explain it.

- a **textarea** attribute

#### You can copy the value of a **Text** attribute to:
- a **textarea** attribute
- a **Simple select** attribute (the option code should exist)
- a **Reference entity single link** attribute *(the record should exist)*

#### You can copy the value of an **Identifier** attribute to:
- a **text** attribute
- a **textarea** attribute

#### You can copy the value of a **Date** attribute to:
- a **text** attribute

> - a **textarea** attribute

Reviewer 2: Can be useful to explain the date format

#### You can copy the value of a **Metric** attribute to:
- a **text** attribute
- a **textarea** attribute
- a **number** attribute

#### You can copy the value of a **Number** attribute to:
- a **text** attribute
- a **textarea** attribute
- a **metric** attribute

#### You can copy the value of a **Price** attribute to:
- a **text** attribute
- a **textarea** attribute

#### You can copy the value of a **Reference entity single link** attribute to:
- a **text** attribute
- a **textarea** attribute
- a **simple select** attribute *(the option code of the simple select attribute should exist)*

#### You can copy the value of a **Reference entity multiple link** attribute to:
- a **text** attribute
- a **textarea** attribute

> - a **multi select** attribute *(the option code of the multi select attribute should exist)*

Reviewer 2: ```suggestion
- a **multi select** attribute *(the option codes of the multi select attribute should exist)*
```

> - a **multi select** attribute *(the option code of the multi select attribute should exist)*

Reviewer 5: ```suggestion
- a **multi select** attribute *(the option code of the multi select attribute must already exist)*
```
